dnadoh/synthesize.py

```diff
@@ -0,0 +1,335 @@
+"""Synthesize data."""
+
+import argparse
+import csv
+import json
+import random
+from typing import List, Optional
+
+from pydantic import BaseModel
+
+# --------------------------------------------------------------------------------------
+# Genomes
+# --------------------------------------------------------------------------------------
+
+
+DNA = "ACGT"
+SNP_PROBS = (0.50, 0.25, 0.13, 0.12)
+
+
+class GenePool(BaseModel):
+    """A set of genomes."""
+
+    # Number of bases in each genome.
+    length: int
+
+    # Reference genome without mutations.
+    reference: str
+
+    # Individual genomes.
+    individuals: List[str]
+
+    # Locations of mutations.
+    locations: List[int]
+
+    def __str__(self):
+        """Printable representation."""
+        return "\n".join(
+            [
+                f"length: {self.length}",
+                f"reference: {self.reference}",
+                "individuals:",
+                "\n".join(f"- {i}" for i in self.individuals),
+                "locations:",
+                "\n".join(f"- {loc}" for loc in self.locations),
+            ]
+        )
+
+
+def random_sequence(length):
+    """Generate a random sequence of bases of the specified length."""
+    assert 0 < length
+    return "".join(random.choices(DNA, k=length))
+
+
+def random_genomes(length, num_genomes, num_snps, max_num_other_mutations):
+    """Generate a set of genomes with specified number of point mutations.
+
+    1.  Create a reference genome.
+    2.  Generate `num_genomes` copies.
+    3.  Pick `num_mutations` distinct locations.
+    4.  Introduce significant mutations using `SNP_PROBS` as weights.
+    5.  Introduce up to `max_num_other_mutations` at other locations (unweighted).
+    """
+    assert 0 <= num_snps <= length
+
+    # Reference genomes and specific genomes to modify.
+    reference = random_sequence(length)
+    individuals = [reference] * num_genomes
+
+    # Locations for SNPs.
+    locations = random.sample(list(range(length)), num_snps)
+
+    # Introduce significant mutations.
+    for loc in locations:
```

Comment:
so these are locations where we will always try to simulate a mutation for?

Reply:
yes